leases/tasks/send_invoices.py

```diff
@@ -0,0 +1,231 @@
+        harbor=lease.berth.pier.harbor
+    ).first():
+        product = harbor_product
+
+    # If there's no Harbor-specific product, try with the "default" product
+    else:
+        product = products_for_pricegroup.filter(harbor__isnull=True).first()
+
+    return product
+
+
+def get_winter_storage_product(lease: WinterStorageLease) -> WinterStorageProduct:
+    raise NotImplementedError
+
+
+def get_valid_leases(season_start: date, lease_type: str) -> QuerySet:
+    if lease_type == BERTH_LEASES:
+        return get_valid_berth_leases(season_start)
+    return get_valid_winter_storage_leases(season_start)
+
+
+def get_valid_winter_storage_leases(season_start: date) -> QuerySet:
+    raise NotImplementedError
+
+
+def get_valid_berth_leases(season_start: date) -> QuerySet:
+    """
+    Get the leases that were active last year
+    If today is:
+      (1) before season: leases from last year
+      (2) during or after season: leases from this year
+    """
+
+    current_date = today().date()
+    # If today is before the season starts but during the same year (1)
+    if current_date < season_start and current_date.year == season_start.year:
+        lease_year = current_date.year - 1
+    else:  # (2)
+        lease_year = current_date.year
+
+    # Filter leases from the upcoming season
+    future_leases = BerthLease.objects.filter(
+        start_date__year__gt=lease_year,
+        berth=OuterRef("berth"),
+        customer=OuterRef("customer"),
+    )
+
+    # Exclude leases that have already been assigned to the same customer and berth on the future
+    leases: QuerySet = BerthLease.objects.exclude(
+        Exists(future_leases.values("pk"))
+    ).filter(
+        # Only allow leases that are auto-renewing and have been paid
+        renew_automatically=True,
+        status=LeaseStatus.PAID,
+        start_date__year=lease_year,
+        end_date__year=lease_year,
+    )
+
+    return leases
+
+
+def create_new_lease(
+    lease: Union[BerthLease, WinterStorageLease], start_date: date, end_date: date
+) -> Union[BerthLease, WinterStorageLease]:
+    """Create a new lease instance"""
+
+    # Blank the PK to signal that a new instance has to be created
+    lease.pk = None
+
+    # Manually need to set creating because of the checks performed on clean
+    lease._state.adding = True
+
+    # Update the dates for the lease for the next season
+    lease.start_date = start_date
+    lease.end_date = end_date
+    lease.save()
+
+    return lease
+
+
+def send_berth_invoices(request: HttpRequest, due_date: date = None) -> Dict[str, List]:
+    return send_invoices(
+        request,
+        BERTH_LEASES,
+        season_start=calculate_season_start_date(),
+        season_end=calculate_season_end_date(),
+        due_date=due_date,
+    )
+
+
+def send_winter_storage_invoices(
+    request: HttpRequest, due_date: date = None
+) -> Dict[str, List]:
+    raise NotImplementedError
+
+
+@transaction.atomic
+def send_invoices(
```

Comment:
Error handling/risk management suggestion: would it make sense to have a "maximum_failures" parameter? If too many leases fail, then throw an exception, which results in the transaction being cancelled and no changes made. That way the databse will not accidentally end up with possibly 1000s of leases in LeaseStatus.ERROR, which probably would require manual cleanup? Maybe set the maximum_failures to something like 50 - that is probably an amount the admins can sort out manually? Or are there some classes of lease failures that are normal and expected?

Reply:
Good point 🤔 worth exploring the idea. I'll see if I can reach a simple version that works for this MVP.